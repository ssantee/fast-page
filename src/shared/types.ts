export type DomainEnv = {
  domain: string;
  account: string;
  region: string;
  adminDomain: string;
};
