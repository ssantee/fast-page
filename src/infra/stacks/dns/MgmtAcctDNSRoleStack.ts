import {
  aws_route53 as route53,
  aws_iam as iam,
  CfnOutput,
  Stack,
  StackProps,
  Duration,
} from "aws-cdk-lib";
import { Construct } from "constructs";
import { DomainEnv } from "../../../shared/types";

interface MgmtAcctDNSProps extends StackProps {
  iamPrincipalAccountNo: string;
  mgmtEnv: DomainEnv;
  targetEnv: DomainEnv;
  apiDomain: string;
}

export class MgmtAcctDNSRoleStack extends Stack {
  constructor(scope: Construct, id: string, props: MgmtAcctDNSProps) {
    super(scope, id);

    const newZone = new route53.PublicHostedZone(this, "HostedZone", {
      zoneName: props.mgmtEnv.domain,
    });

    const newAdminZone = new route53.PublicHostedZone(this, "HostedZoneAdmin", {
      zoneName: props.mgmtEnv.adminDomain,
    });

    const newApiZone = new route53.PublicHostedZone(this, "HostedZoneApi", {
      zoneName: props.apiDomain,
    });

    const crossAccountRole = new iam.Role(this, "CrossAccountRole", {
      // The role name must be predictable
      roleName: "MyXAcctDelegationRole",
      // The other account
      assumedBy: new iam.AccountPrincipal(props.targetEnv.account),
      // You can scope down this role policy to be least privileged.
      // If you want the other account to be able to manage specific records,
      // you can scope down by resource and/or normalized record names
      inlinePolicies: {
        crossAccountPolicy: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              sid: "ListHostedZonesByName",
              effect: iam.Effect.ALLOW,
              actions: ["route53:ListHostedZonesByName"],
              resources: ["*"],
            }),
            new iam.PolicyStatement({
              sid: "GetHostedZoneAndChangeResourceRecordSets",
              effect: iam.Effect.ALLOW,
              actions: [
                "route53:GetHostedZone",
                "route53:ChangeResourceRecordSets",
              ],
              // This example assumes the RecordSet subdomain.somexample.com
              // is contained in the HostedZone
              resources: [
                `arn:aws:route53:::hostedzone/${newZone.hostedZoneId}`,
              ],
              conditions: {
                "ForAllValues:StringLike": {
                  "route53:ChangeResourceRecordSetsNormalizedRecordNames": [
                    "dev.fp.santee.cloud",
                    "prod.fp.santee.cloud",
                    "fp.santee.cloud",
                    "fpadmin.santee.cloud",
                    "dev.fpadmin.santee.cloud",
                    "prod.fpadmin.santee.cloud",
                    "fpapi.santee.cloud",
                    "prod.fpapi.santee.cloud",
                    "dev.fpapi.santee.cloud",
                  ],
                },
              },
            }),
          ],
        }),
      },
    });
    newZone.grantDelegation(crossAccountRole);
    newAdminZone.grantDelegation(crossAccountRole);
    newApiZone.grantDelegation(crossAccountRole);

    const tldZone = route53.PublicHostedZone.fromPublicHostedZoneAttributes(
      this,
      "TLDZone",
      {
        hostedZoneId: "Z02022183L2XHLGIP6A83",
        zoneName: "santee.cloud",
      },
    );

    new route53.NsRecord(this, "SubDomainNSRecord", {
      zone: tldZone,
      recordName: props.mgmtEnv.domain,
      values: newZone.hostedZoneNameServers || [],
      ttl: Duration.seconds(172800),
    });

    new route53.NsRecord(this, "DomainNSRecordAdmin", {
      zone: tldZone,
      recordName: props.mgmtEnv.adminDomain,
      values: newAdminZone.hostedZoneNameServers || [],
      ttl: Duration.seconds(172800),
    });

    new route53.NsRecord(this, "ApiDomainNSRecord", {
      zone: tldZone,
      recordName: props.apiDomain,
      values: newApiZone.hostedZoneNameServers || [],
      ttl: Duration.seconds(172800),
    });

    new CfnOutput(this, "CrossAccountDelegationRoleArn", {
      description: "CrossAccountDelegationRoleArn",
      value: crossAccountRole.roleArn,
    });
  }
}
