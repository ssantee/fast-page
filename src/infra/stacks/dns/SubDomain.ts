import {
  aws_route53 as route53,
  aws_iam as iam,
  aws_ssm as ssm,
  Stack,
  StackProps,
  SecretValue,
  RemovalPolicy,
} from "aws-cdk-lib";
import { Construct } from "constructs";
import { DomainEnv } from "../../../shared/types";
import {
  Certificate,
  CertificateValidation,
} from "aws-cdk-lib/aws-certificatemanager";
import { Secret } from "aws-cdk-lib/aws-secretsmanager";

interface SubDomainProps extends StackProps {
  mgmtEnvAcctNo: string;
  certificateArnParamName: string;
  subdomainHostedZoneIdParamName: string;
  deployEnvDomain: string;
  mgmtEnvDomain: string;
}

export class SubDomain extends Stack {
  constructor(scope: Construct, id: string, props: SubDomainProps) {
    super(scope, id);

    const subZone = new route53.PublicHostedZone(this, "Route53FPSubZone", {
      zoneName: props.deployEnvDomain,
    });

    // import the delegation role by constructing the roleArn
    const delegationRoleArn = Stack.of(this).formatArn({
      region: "", // IAM is global in each partition
      service: "iam",
      account: props.mgmtEnvAcctNo,
      resource: "role",
      resourceName: "MyXAcctDelegationRole", // Do not change
    });

    const delegationRole = iam.Role.fromRoleArn(
      this,
      "DelegationRole",
      delegationRoleArn,
    );

    // create the record
    new route53.CrossAccountZoneDelegationRecord(
      this,
      "CrossAccountZoneDelegationRecord",
      {
        delegatedZone: subZone,
        parentHostedZoneName: props.mgmtEnvDomain,
        delegationRole,
        assumeRoleRegion: "us-east-1",
      },
    );

    const certificate = new Certificate(this, "DelegatedZoneCert", {
      domainName: props.deployEnvDomain,
      validation: CertificateValidation.fromDns(subZone),
    });

    new ssm.StringParameter(this, "mySsmParameter", {
      parameterName: props.certificateArnParamName,
      stringValue: certificate.certificateArn,
    });

    new ssm.StringParameter(this, "mySsmParameter1", {
      parameterName: props.subdomainHostedZoneIdParamName,
      stringValue: subZone.hostedZoneId,
    });
  }
}
