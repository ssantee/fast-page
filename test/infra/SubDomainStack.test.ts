import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { MgmtAcctDNSRoleStack } from "../../src/infra/stacks/dns/MgmtAcctDNSRoleStack";
import { describe } from "node:test";
import { DomainEnv } from "../../src/shared/types";
import { SubDomain } from "../../src/infra/stacks/dns/SubDomain";

describe("SubDomainStack test suite", () => {
  let SubDomainStackTemplate: Template;
  const envs: { [env: string]: DomainEnv } = {
    dev: {
      account: "123456789012",
      region: "us-east-1",
      domain: "dev.pg.santee.cloud",
      adminDomain: "dev.pgadmin.santee.cloud",
      apiDomain: "dev.pgapi.santee.cloud",
    },
    prod: {
      account: "123456789012",
      region: "us-east-1",
      domain: "prod.pg.santee.cloud",
      adminDomain: "dev.pgadmin.santee.cloud",
      apiDomain: "prod.pgapi.santee.cloud",
    },
    root: {
      account: "123456789012",
      region: "us-east-1",
      domain: "pg.santee.cloud",
      adminDomain: "pgadmin.santee.cloud",
      apiDomain: "pgapi.santee.cloud",
    },
  };

  beforeAll(() => {
    const app = new cdk.App();
    const subDomainStack = new SubDomain(app, "WebPublicStack", {
      env: {
        account: envs.prod.account,
        region: "us-east-1",
      },
      certificateArnParamName: "/account/fpCertificateArn",
      subdomainHostedZoneIdParamName: "/account/fpSubdomainHostedZoneId",
      mgmtEnvAcctNo: envs.root.account,
      deployEnvDomain: envs.prod.domain,
      mgmtEnvDomain: envs.root.domain,
    });
    SubDomainStackTemplate = Template.fromStack(subDomainStack);
  });

  test("creates cross account zone delegation record", () => {
    SubDomainStackTemplate.hasResourceProperties(
      "Custom::CrossAccountZoneDelegation",
      {
        ParentZoneName: envs.root.domain,
        DelegatedZoneName: envs.prod.domain,
        AssumeRoleArn: {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                Ref: "AWS::Partition",
              },
              `:iam::${envs.root.account}:role/MyXAcctDelegationRole`,
            ],
          ],
        },
      },
    );
  });

  test("creates hosted zone", () => {
    SubDomainStackTemplate.hasResourceProperties("AWS::Route53::HostedZone", {
      Name: `${envs.prod.domain}.`,
    });
  });

  test("Matches snapshot", () => {
    expect(SubDomainStackTemplate).toMatchSnapshot();
  });
});
