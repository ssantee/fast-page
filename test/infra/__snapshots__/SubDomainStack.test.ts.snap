// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CrossAccountZoneDelegationRecordCrossAccountZoneDelegationCustomResourceE78B0799": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CrossAccountZoneDelegationRecordcrossaccountzonedelegationhandlerrolePolicy7D87EC56",
      ],
      "Properties": {
        "AssumeRoleArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::123456789012:role/MyXAcctDelegationRole",
            ],
          ],
        },
        "AssumeRoleRegion": "us-east-1",
        "DelegatedZoneName": "prod.pg.santee.cloud",
        "DelegatedZoneNameServers": {
          "Fn::GetAtt": [
            "Route53FPSubZoneFFAE6E0D",
            "NameServers",
          ],
        },
        "ParentZoneName": "pg.santee.cloud",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCrossAccountZoneDelegationCustomResourceProviderHandler44A84265",
            "Arn",
          ],
        },
        "TTL": 172800,
      },
      "Type": "Custom::CrossAccountZoneDelegation",
      "UpdateReplacePolicy": "Delete",
    },
    "CrossAccountZoneDelegationRecordcrossaccountzonedelegationhandlerrolePolicy7D87EC56": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iam::123456789012:role/MyXAcctDelegationRole",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CrossAccountZoneDelegationRecordcrossaccountzonedelegationhandlerrolePolicy7D87EC56",
        "Roles": [
          {
            "Fn::Select": [
              1,
              {
                "Fn::Split": [
                  "/",
                  {
                    "Fn::Select": [
                      5,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomCrossAccountZoneDelegationCustomResourceProviderHandler44A84265": {
      "DependsOn": [
        "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "85d541e10871ed2e6dced66cbf72398d8833cfb715b66529e614fac17364e800.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DelegatedZoneCert71201BFD": {
      "Properties": {
        "DomainName": "prod.pg.santee.cloud",
        "DomainValidationOptions": [
          {
            "DomainName": "prod.pg.santee.cloud",
            "HostedZoneId": {
              "Ref": "Route53FPSubZoneFFAE6E0D",
            },
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "WebPublicStack/DelegatedZoneCert",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "Route53FPSubZoneFFAE6E0D": {
      "Properties": {
        "Name": "prod.pg.santee.cloud.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "mySsmParameter1B506219F": {
      "Properties": {
        "Name": "/account/fpSubdomainHostedZoneId",
        "Type": "String",
        "Value": {
          "Ref": "Route53FPSubZoneFFAE6E0D",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "mySsmParameter72A709A2": {
      "Properties": {
        "Name": "/account/fpCertificateArn",
        "Type": "String",
        "Value": {
          "Ref": "DelegatedZoneCert71201BFD",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
